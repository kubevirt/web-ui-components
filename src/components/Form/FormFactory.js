import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { FormGroup, Col, ControlLabel, HelpBlock, Form, FieldLevelHelp } from 'patternfly-react';
import { get } from 'lodash';

import { Dropdown, Checkbox, Text, TextArea, Integer } from '.';
import { VALIDATION_INFO_TYPE } from '../../constants';
import { getValidationObject } from '../../utils/validations';
import { ERROR_IS_REQUIRED } from '../Wizard/CreateVmWizard/strings';

export const getFormElement = props => {
  const {
    type,
    id,
    value,
    title,
    onChange,
    onBlur,
    choices,
    defaultValue,
    isControlled,
    disabled,
    className,
    LoadingComponent,
  } = props;
  switch (type) {
    case 'textarea':
      return (
        <TextArea
          id={id}
          key={id}
          value={isControlled ? value || '' : undefined}
          defaultValue={isControlled ? undefined : defaultValue}
          onBlur={onBlur}
          onChange={onChange}
          disabled={disabled}
        />
      );
    case 'dropdown':
      return (
        <Dropdown
          id={id}
          value={value || defaultValue}
          onChange={onChange}
          onBlur={onBlur}
          choices={choices}
          disabled={disabled}
        />
      );
    case 'checkbox':
      return (
        <Checkbox
          id={id}
          key={id}
          title={title}
          checked={isControlled ? value || false : undefined}
          onBlur={onBlur}
          onChange={onChange}
          disabled={disabled}
        />
      );
    case 'positive-number':
      return (
        <Integer
          id={id}
          key={id}
          value={isControlled ? value || '' : undefined}
          defaultValue={isControlled ? undefined : defaultValue}
          onBlur={onBlur}
          onChange={onChange}
          positive
          disabled={disabled}
        />
      );
    case 'label':
      return (
        <div className={className} key={id}>
          {value}
        </div>
      );
    case 'loading':
      return <LoadingComponent />;
    default:
      return (
        <Text
          id={id}
          key={id}
          value={isControlled ? value || '' : undefined}
          defaultValue={isControlled ? undefined : defaultValue}
          onBlur={onBlur}
          onChange={onChange}
          disabled={disabled}
        />
      );
  }
};

const onChange = (field, value, key, onFormChange) => {
  let validation;

  if (field.required && String(value).trim().length === 0) {
    validation = getValidationObject(ERROR_IS_REQUIRED);
  } else if (field.validate) {
    validation = field.validate(value);
  }

  if (validation) {
    validation.message = `${field.title} ${validation.message}`;
  }

  onFormChange({ value, validation }, key);
};

const getFormGroups = ({ fields, fieldsValues, onFormChange, textPosition, labelSize, controlSize, horizontal }) =>
  Object.keys(fields)
    .filter(key => fields[key] && (!fields[key].isVisible || fields[key].isVisible(fieldsValues)))
    .map(key => {
      const field = fields[key];
      const values = fieldsValues[key];
      const validation = get(values, 'validation');
      const value = get(values, 'value');
      const hasValidationMessage = !!get(validation, 'message');

      const child = getFormElement({
        ...field,
        value,
        isControlled: true,
        onChange: newValue => onChange(field, newValue, key, onFormChange),
      });

      const label = horizontal &&
        field.title && (
          <Col sm={labelSize} className={textPosition}>
            {field.type !== 'checkbox' && (
              <React.Fragment>
                <ControlLabel className={field.required ? 'required-pf' : null}>{field.title}</ControlLabel>
                {field.help && (
                  <FieldLevelHelp className="kubevirt-form-group__field-help" placement="right" content={field.help} />
                )}
              </React.Fragment>
            )}
          </Col>
        );

      return (
        <FormGroup
          key={key}
          validationState={validation && validation.type !== VALIDATION_INFO_TYPE ? validation.type : null}
          className={classNames('kubevirt-form-group', {
            'kubevirt-form-group--no-bottom': horizontal && field.noBottom,
            'kubevirt-form-group--help': !horizontal && hasValidationMessage,
            'kubevirt-form-group--no-help': !horizontal && !hasValidationMessage,
          })}
        >
          {label}
          <Col sm={controlSize}>
            {child}
            <HelpBlock>{get(validation, 'message')}</HelpBlock>
          </Col>
        </FormGroup>
      );
    });

export const InlineFormFactory = ({ fields, fieldsValues, onFormChange }) =>
  getFormGroups({ fields, fieldsValues, onFormChange, horizontal: true });

export const ListFormFactory = ({ fields, fieldsValues, onFormChange, actions, columnSizes }) => {
  const formGroups = getFormGroups({ fields, fieldsValues, onFormChange });
  const form = formGroups.map((formGroup, index) => (
    <Col
      key={`col-${index}`}
      {...columnSizes}
      className={classNames('kubevirt-list-form-factory__column', {
        'kubevirt-list-form-factory__column--last': index === formGroups.length - 1,
      })}
    >
      {formGroup}
    </Col>
  ));

  return (
    <React.Fragment>
      {form}
      <div className="kubevirt-list-form-factory__actions">{actions}</div>
    </React.Fragment>
  );
};

ListFormFactory.propTypes = {
  fields: PropTypes.object.isRequired,
  fieldsValues: PropTypes.object.isRequired,
  onFormChange: PropTypes.func.isRequired,
  actions: PropTypes.object.isRequired,
  columnSizes: PropTypes.object.isRequired,
};

export const FormFactory = ({
  fields,
  fieldsValues,
  onFormChange,
  textPosition,
  labelSize,
  controlSize,
  formClassName,
}) => {
  const formGroups = getFormGroups({
    fields,
    fieldsValues,
    onFormChange,
    textPosition,
    labelSize,
    controlSize,
    horizontal: true,
  });
  return (
    <Form horizontal className={formClassName}>
      {formGroups}
    </Form>
  );
};

FormFactory.defaultProps = {
  textPosition: 'text-right',
  labelSize: 3,
  controlSize: 5,
  formClassName: undefined,
};

FormFactory.propTypes = {
  fields: PropTypes.object.isRequired,
  fieldsValues: PropTypes.object.isRequired,
  onFormChange: PropTypes.func.isRequired,
  textPosition: PropTypes.string,
  labelSize: PropTypes.number,
  controlSize: PropTypes.number,
  formClassName: PropTypes.string,
};
