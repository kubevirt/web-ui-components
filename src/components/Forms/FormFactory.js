import React from 'react';
import PropTypes from 'prop-types';
import { FormGroup, Col, ControlLabel, HelpBlock, Form, FieldLevelHelp } from 'patternfly-react';
import { get, has } from 'lodash';
import { TextArea, Dropdown, Checkbox, Text } from '.';

export const FormFactory = ({ fields, fieldsValues, onFormChange }) => {
  const formGroups = Object.keys(fields)
    .filter(key => !fields[key].isVisible || fields[key].isVisible(fieldsValues))
    .map(key => {
      let child;
      const validMsg = get(fieldsValues[key], 'validMsg');
      switch (fields[key].type) {
        case 'textarea':
          child = <TextArea fieldKey={key} value={get(fieldsValues[key], 'value', '')} onChange={onFormChange} />;
          break;
        case 'dropdown':
          child = (
            <Dropdown
              fieldKey={key}
              value={has(fieldsValues[key], 'value') ? fieldsValues[key].value : fields[key].default}
              choices={typeof fields[key].values === 'function' ? fields[key].values() : fields[key].values}
              onChange={onFormChange}
            />
          );
          break;
        case 'checkbox':
          child = (
            <Checkbox
              fieldKey={key}
              checked={get(fieldsValues[key], 'value', false)}
              title={fields[key].title}
              onChange={onFormChange}
            />
          );
          break;
        case 'text':
        default:
          child = <Text fieldKey={key} value={get(fieldsValues[key], 'value', '')} onChange={onFormChange} />;
      }
      return (
        <FormGroup
          key={key}
          validationState={validMsg ? 'error' : null}
          className={fields[key].noBottom ? 'form-group-no-bottom' : undefined}
        >
          <Col sm={3} className="text-right">
            {fields[key].type === 'checkbox' ? null : (
              <React.Fragment>
                <ControlLabel className={fields[key].required ? 'required-pf' : null}>{fields[key].title}</ControlLabel>
                {fields[key].help ? <FieldLevelHelp placement="right" content={fields[key].help()} /> : undefined}
              </React.Fragment>
            )}
          </Col>
          <Col sm={5}>
            {child}
            <HelpBlock>{validMsg}</HelpBlock>
          </Col>
        </FormGroup>
      );
    });
  return <Form horizontal>{formGroups}</Form>;
};

FormFactory.propTypes = {
  fields: PropTypes.object.isRequired,
  fieldsValues: PropTypes.object.isRequired,
  onFormChange: PropTypes.func.isRequired
};
