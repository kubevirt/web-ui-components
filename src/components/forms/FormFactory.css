.form-group-no-bottom {
  margin-bottom: 2px;
}

.popover {
  max-width: 400px;
}

