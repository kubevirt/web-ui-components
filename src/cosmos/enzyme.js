import createContext from 'react-cosmos-test/enzyme';

import paths from '../../config/paths';

export default function createTestContext(args) {
  args.proxies = args.proxies || []; // workaround of react-cosmos-test bug
  return createContext({
    ...args,
    cosmosConfigPath: `${paths.config}/cosmos.config.js`,
  });
}
