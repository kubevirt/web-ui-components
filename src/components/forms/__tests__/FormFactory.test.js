import React from 'react';
import { shallow } from 'enzyme';
import { FormFactory } from '../FormFactory';
import { minFormFields, formFieldsValues, allFormFields } from '../FormFactory.fixtures';

const testCreateFormFactory = () => (
  <FormFactory fields={minFormFields} fieldsValues={formFieldsValues} onFormChange={() => {}} />
);

const testCreateFormFactoryAllFields = () => (
  <FormFactory fields={allFormFields} fieldsValues={formFieldsValues} onFormChange={() => {}} />
);

describe('<FormFactory />', () => {
  it('renders correctly', () => {
    const component = shallow(testCreateFormFactory());
    expect(component).toMatchSnapshot();
  });
  it('required field is marked', () => {
    const component = shallow(testCreateFormFactoryAllFields());
    expect(component.find('.required-pf')).toHaveLength(1);
  });
  it('not visible field is not rendered', () => {
    const component = shallow(testCreateFormFactory());
    expect(component.find('.control-label').find(node => node.text() === 'invisibleField')).toHaveLength(0);
  });
});
