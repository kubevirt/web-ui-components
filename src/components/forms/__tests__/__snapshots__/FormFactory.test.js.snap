// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormFactory /> renders correctly 1`] = `
<Form
  bsClass="form"
  componentClass="form"
  horizontal={true}
  inline={false}
>
  <FormGroup
    bsClass="form-group"
    key="textField"
    validationState="error"
  >
    <Col
      bsClass="col"
      className="text-right"
      componentClass="div"
      sm={3}
    >
      <ControlLabel
        bsClass="control-label"
        className={null}
        srOnly={false}
      >
        textField
      </ControlLabel>
    </Col>
    <Col
      bsClass="col"
      componentClass="div"
      sm={5}
    >
      <Component
        fieldKey="textField"
        onChange={[Function]}
        value="textFieldValue"
      />
      <HelpBlock
        bsClass="help-block"
      >
        validation message
      </HelpBlock>
    </Col>
  </FormGroup>
</Form>
`;
